flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java

```diff
@@ -207,6 +210,40 @@ public void testRestClientClosedHandling() throws Exception {
         }
     }
 
+    /**
+     * Tests that the futures returned by {@link RestClient} fail immediately if the client is
+     * already closed.
+     *
+     * <p>See FLINK-32583
+     */
+    @Test
+    public void testCloseClientBeforeRequest() throws Exception {
```

Comment:
I looked into the multi-threading of netty a bit more and found the `SelectStrategy` a viable tool to control the state of the request. I came up with the following test:
```
@Test
    public void testCloseClientWhileProcessingRequest() throws Exception {
        final OneShotLatch connectTriggered = new OneShotLatch();
        final OneShotLatch closeTriggered = new OneShotLatch();
        final SelectStrategy fallbackSelectStrategy =
                DefaultSelectStrategyFactory.INSTANCE.newSelectStrategy();
        final SelectStrategy selectStrategy =
                (selectSupplier, hasTasks) -> {
                    connectTriggered.trigger();
                    closeTriggered.awaitQuietly();

                    return fallbackSelectStrategy.calculateStrategy(selectSupplier, hasTasks);
                };
        // Note that the executor passed to the RestClient constructor is not the same as the
        // executor used by Netty
        try (final RestClient restClient =
                new RestClient(
                        new Configuration(), Executors.directExecutor(), () -> selectStrategy)) {
            final CompletableFuture<?> requestFuture =
                    restClient.sendRequest(
                            unroutableIp,
                            80,
                            new TestMessageHeaders(),
                            EmptyMessageParameters.getInstance(),
                            EmptyRequestBody.getInstance());

            connectTriggered.await();

            final CompletableFuture<Void> closeFuture = restClient.closeAsync();

            closeTriggered.trigger();

            // close should complete successfully
            closeFuture.get();

            final Throwable cause =
                    assertThrows(
                                    ExecutionException.class,
                                    () -> requestFuture.get(0, TimeUnit.SECONDS))
                            .getCause();
            assertThat(cause, instanceOf(IllegalStateException.class));
        }
    }
```
It would require an extension of the `RestClient` adding a package-private constructor:
```
    // ...
    public RestClient(Configuration configuration, Executor executor)
            throws ConfigurationException {
        this(configuration, executor, DefaultSelectStrategyFactory.INSTANCE);
    }

    @VisibleForTesting
    RestClient(
            Configuration configuration,
            Executor executor,
            SelectStrategyFactory selectStrategyFactory)
    //...
    NioEventLoopGroup group =
                new NioEventLoopGroup(
                        1,
                        new ExecutorThreadFactory("flink-rest-client-netty"),
                        SelectorProvider.provider(),
                        selectStrategyFactory);
    // ...
```
WDYT? That should test the `isRunning.get()` code path.

Reply:
Btw. in that case, the future is failing with an `IllegalStateException`. I'm wondering whether we should switch to that one within the `RestClient` as well to make it consistent. Or we wrap the `IllegalStateException` in an `IOException` because I think that `IOException` still seems to be the best fit. :thinking:

Reply:
Cool, good find, I'll try it out.

Regarding the exception, I've started to feel that maybe `IOException` actually _isn't_ the best choice, since no true IO (e.g. via syscalls) is happening at the time of the failure. Even if we take an analogous example of trying to write to a file that's closed, the error would often come from the failure of the syscall to write.

In this situation, the client is in fact in an illegal state (closed) for making requests.

Reply:
With this additional test, the `isRunning` codepath is now tested, but not the branch where the listener is unable to be attached and is thus handled in `notifyResponseFuturesOfShutdown()`. I think there is just a microscopic window for this to happen, between the calls to `bootstrap.connect()` and `connectFuture.addListener()`, which are not executed atomically.

The only way that comes to mind to test that would be to inject a latch to enable calling `close()` between those two calls.

Do you think that's necessary? I could also add a note that that eventuality was tested manually, with `notifyResponseFuturesOfShutdown()` successfully resolving the future.

Reply:
hm, you're right - that's a tricky one. We could extend the `RestClient` constructor that's labeled as `@VisibleForTesting` to not only provide the `SelectyStrategyFactory` but also a `Collection` to initialize the `responseChannelFutures` field. The default value would be `ConcurrentHashMap.newKeySet()`.

That would enable us to check in `testCloseClientWhileProcessingRequest` that the future is, indeed, added to the collection when calling connect (i.e. when executing the custom `SelectStrategy` callback) and cleared afterwards.

And then, we could add another test that adds a non-empty collection via the constructor to the `RestClient` and calls `closeAsync` on that client to verify that the future was completed exceptionally and the collection cleared afterwards. :thinking: 

Maybe, I got too nitpicky on that one :innocent: WDYT?

Reply:
Might be able to get the same behavior with just a `@VisibleForTesting` getter for `responseChannelFutures`, I'll give it a shot.

Always makes me uneasy allowing injection or adding extra steps to initializing `ConcurrentXyz` fields, even for testing, since you can't enforce the concurrency-safety with the type system.

Reply:
that's a fair point - I know what you mean. :-D